src/main/java/com/squareup/javapoet/TypeName.java

```diff
@@ -87,14 +88,34 @@
 
   /** The name of this type if it is a keyword, or null. */
   private final String keyword;
+  public final List<AnnotationSpec> annotations;
 
   private TypeName(String keyword) {
+    this(keyword, new ArrayList<AnnotationSpec>());
+  }
+
+  private TypeName(String keyword, List<AnnotationSpec> annotations) {
     this.keyword = keyword;
+    this.annotations = Util.immutableList(annotations);
   }
 
   // Package-private constructor to prevent third-party subclasses.
-  TypeName() {
-    this(null);
+  TypeName(List<AnnotationSpec> annotations) {
+    this(null, annotations);
+  }
+
+  public TypeName annotated(AnnotationSpec... annotations) {
+    Util.checkNotNull(annotations, "annotations == null");
+    return new TypeName(keyword, Arrays.asList(annotations));
```

Comment:
This replaces any existing annotations with the new collection. If the user wants to merge the old and new annotations, it has to be done before calling this method. A nice side-effect of this solution, that an empty new collection strips any existing annotations.